### ISSUE TITLE
opt: sqllite logic test optimizer verification errors

### ISSUE THREAD

**Issue by cucaroach** [cockroachdb/cockroach#78615]
$ ./dev test ./pkg/sql/logictest -v -f=TestSqlLiteLogic/fakedist-disk/slt_good_0.test --test-args '-bigtest'

```
        /Users/tpr/go/src/github.com/cockroachdb/sqllogictest/test/index/between/1/slt_good_0.test:437: SELECT pk FROM tab1 WHERE (col4 IN (5.85,5.95,6.30,0.0,1.27) AND (col4 IN (9.41,4.56,2.55,0.83,3.95,6.92) AND (col1 = 6.24)) OR col0 IN (SELECT col3 FROM tab1 WHERE (((((col0 IN (8,1) AND col0 IS NULL AND col1 < 9.21) AND col4 > 8.64 AND (col3 >= 9) OR col0 < 5)))))) AND (col1 < 7.95 OR col0 < 0) AND col3 >= 9 AND col0 < 2 OR col4 BETWEEN 1.13 AND 0.71 AND (col3 > 9) AND (col0 < 3)
        expected success, but found
        (XX000) internal error: output cols mismatch: (64-66,68,69) vs (14,64-66,68,69)
        verify.go:89: in VerifyAgainst()
        DETAIL: stack trace:
        github.com/cockroachdb/cockroach/pkg/sql/opt/props/verify.go:89: VerifyAgainst()
        github.com/cockroachdb/cockroach/pkg/sql/opt/memo/check_expr.go:54: CheckExpr()
        github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:24006: AddLookupJoinToGroup()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/join_funcs.go:758: func1()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/scan_index_iter.go:298: ForEachStartingAfter()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/scan_index_iter.go:209: ForEach()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/join_funcs.go:359: generateLookupJoinsImpl()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/join_funcs.go:256: GenerateLookupJoins()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:1865: exploreSemiJoin()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:34: exploreGroupMember()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.go:182: exploreGroup()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:495: optimizeGroup()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:265: optimizeExpr()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:537: optimizeGroupMember()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:482: optimizeGroup()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:265: optimizeExpr()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:537: optimizeGroupMember()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:482: optimizeGroup()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:265: optimizeExpr()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:537: optimizeGroupMember()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:482: optimizeGroup()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:265: optimizeExpr()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:537: optimizeGroupMember()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:482: optimizeGroup()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:265: optimizeExpr()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:537: optimizeGroupMember()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:482: optimizeGroup()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:653: optimizeEnforcer()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:612: enforceProps()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:522: optimizeGroupMember()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:482: optimizeGroup()
        github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: Optimize()
```
Epic: CRDB-14914
Jira issue: CRDB-14197

**Comment by cucaroach** [comment]
opt test repro:

```
exec-ddl
CREATE TABLE tab1(pk INTEGER PRIMARY KEY, col0 INTEGER, col1 FLOAT, col2 TEXT, col3 INTEGER, col4 FLOAT, col5 TEXT)
----

exec-ddl
CREATE INDEX idx_tab1_0 on tab1 (col0)
----

exec-ddl
CREATE INDEX idx_tab1_1 on tab1 (col1)
----

exec-ddl
CREATE INDEX idx_tab1_3 on tab1 (col3)
----

exec-ddl
CREATE INDEX idx_tab1_4 on tab1 (col4)
----

opt
SELECT pk FROM tab1 WHERE col1 > 3356.8 AND col4 IN (SELECT col1 FROM tab1 WHERE (col3 = 6354 AND col3 >= 4129 OR col0 >= 5009 AND (col3 > 964) AND (col4 > 4072.6)) AND col0 BETWEEN 3729 AND 8439 AND col0 >= 4860 OR ((col4 < 2879.73)) OR col1 >= 4543.22 OR col3 < 5765)
----
```

**Comment by cucaroach** [comment]
The error stems from here:

https://github.com/cockroachdb/cockroach/blob/ef588e75000fe776296b75192bf7408d39874747/pkg/sql/opt/xform/join_funcs.go#L754

In my test case indexJoin.On.OuterCols adds a column that the inputProps builder in the verification step doesn't so you get (1,3,6) != (1,3,6,12) where 6 = 12 is the join condition.   I also verified that this doesn't occur in the code where this was added so something changed that caused this error to start happening.

**Comment by cucaroach** [comment]
This doesn't repro in 21.2.

**Comment by cucaroach** [comment]
This reproduces 22.1.   Starts with this change:

```
commit cc07c86ad258431de685e8f47e229c2101630bf6 (HEAD)
Author: sumeerbhola <sumeer@cockroachlabs.com>
Date:   Wed Dec 23 16:01:15 2020 -0500

    opt,sql: use paired-joins with non-covering indexes for left joins
    
    This is done when the left outer/semi/anti join can use a
    lookup join. Prior to this, when the non-covering index
    could not fully evaluate the filter for left join we could
    not generate a lookup join.
    
    With this change:
    - Left outer join becomes a pair of two left outer joins.
    - Left semi join is a pair of inner join followed by left
      semi join.
    - Left anti join is a pair of left outer join followed by
      left anti join.
    
    Informs #55452
    
    Release note (performance improvement): The optimizer can now
    generate lookup joins in certain cases for non-covering
    indexes, when performing a left outer/semi/anti join.
```

**Comment by cucaroach** [comment]
Weird that this bug doesn't happen with optsteps but does with just opt and exploretrace.   From exploretrace I can see the problem occurs here:

```
================================================================================
GenerateLookupJoinsWithFilter
================================================================================
Source expression:
  project
   ├── columns: pk:1!null
   └── semi-join (hash)
        ├── columns: pk:1!null col1:3!null col4:6
        ├── select
        │    ├── columns: pk:1!null col1:3!null col4:6
        │    ├── scan tab1
        │    │    └── columns: pk:1!null col1:3 col4:6
        │    └── filters
        │         └── col1:3 > 3356.8
        ├── select
        │    ├── columns: col0:11 col1:12 col3:14 col4:15
        │    ├── scan tab1
        │    │    └── columns: col0:11 col1:12 col3:14 col4:15
        │    └── filters
        │         └── ((((((((col3:14 = 6354) AND (col3:14 >= 4129)) OR (((col0:11 >= 5009) AND (col3:14 > 964)) AND (col4:15 > 4072.6))) AND (col0:11 >= 3729)) AND (col0:11 <= 8439)) AND (col0:11 >= 4860)) OR (col4:15 < 2879.73)) OR (col1:12 >= 4543.22)) OR (col3:14 < 5765)
        └── filters
             └── col4:6 = col1:12

New expression 1 of 1:
  project
   ├── columns: pk:1!null
   └── semi-join (lookup tab1)
        ├── columns: pk:1!null col1:3!null col4:6
        ├── key columns: [10] = [10]
        ├── lookup columns are key
        ├── second join in paired joiner
        ├── inner-join (lookup tab1@idx_tab1_1)
        │    ├── columns: pk:1!null col1:3!null col4:6!null pk:10!null col1:12!null continuation:91
        │    ├── key columns: [6] = [12]
        │    ├── first join in paired joiner; continuation column: continuation:91
        │    ├── select
        │    │    ├── columns: pk:1!null col1:3!null col4:6
        │    │    ├── scan tab1
        │    │    │    └── columns: pk:1!null col1:3 col4:6
        │    │    └── filters
        │    │         └── col1:3 > 3356.8
        │    └── filters (true)
        └── filters
             └── ((((((((col3:14 = 6354) AND (col3:14 >= 4129)) OR (((col0:11 >= 5009) AND (col3:14 > 964)) AND (col4:15 > 4072.6))) AND (col0:11 >= 3729)) AND (col0:11 <= 8439)) AND (col0:11 >= 4860)) OR (col4:15 < 2879.73)) OR (col1:12 >= 4543.22)) OR (col3:14 < 5765)
```

So the key columns: [6] = [12] is the problem.   The indexJoin.On.OuterCols has the 12 and the inputProps.OutputCols doesn't.

**Comment by cucaroach** [comment]
Actually things can be reduced a bit more:

```
exec-ddl
CREATE TABLE tab1(pk INTEGER PRIMARY KEY, col1 FLOAT, col4 FLOAT)
----

exec-ddl
CREATE INDEX idx_tab1_1 on tab1 (col1)
----

exploretrace
SELECT
    pk
FROM
    tab1
WHERE
    col1 > 3356.8
    AND col4 IN (
        SELECT
            col1
        FROM
            tab1
        WHERE
            col4 < 2879.73
            OR col1 >= 4543.22
    )
----
```

**Comment by cucaroach** [comment]
I looked at this for a bit but don't really get it, I think maybe the generated expression doesn't make any sense and we're just missing a check to avoid it but I'm not sure what that check is.  @mgartner @rytaft someone more experienced with optimizer should look at this.

**Comment by mgartner** [comment]
I'll take a look.

**Comment by mgartner** [comment]
This is not a release blocker. I’m removing the labels for now, and will provide a bit more explanation when I get a moment.

**Comment by mgartner** [comment]
I'm fairly certain this is not a release or GA blocker. The issue is that during optimizer test-build assertions we don't correctly calculate the output columns of a semi/anti-join when it is the second join in a paired joiner. Paired joiners break the previous assumption that a semi/anti-join will output all the columns of its input. The second join in a paired join will project away the continuation column, and only output the columns of the first join's input.

**Comment by mgartner** [comment]
This was fixed by #82054. So it turns out this was a release-blocker (or at least a symptom of a release-blocker). When I observed the behavior here, the connection to a correctness bug wasn't obvious. I should have taken the assertion failure more seriously.
